### DISCUSSION TITLE
Calibrating individual circuits before permanent install?

### DISCUSSION THREAD

**Discussion by Overwatching** [emporia-vue-local/esphome#110] [Q&A] [answered]
After a bit of initial confusion, I have the device flashed and I can get readings.

I currently have the device set up on a single phase, with the Phase A clamp and all 16 circuit clamps checking the same circuit.
This circuit has a resistive heater fan attached to it, and it's all going through a Kill-A-Watt which I have measured to be off by +/- 0.25V.
I've done this to see how close the values for each circuit sensor are.

And they are quite a bit farther off than I was expecting.

Kill-A-Watt and Phase A clamp agree on 1375W.
But the individual circuits range from 1316-1421W.

Each [sensor+plug] is consistent though, so it seems like I could calibrate this problem away.

### What is the proper way to do this?

Should I be putting a multiplier in each line of `ct_clamps`?
ex:
`- { phase_id: phase_a, input:  "5", power: { name:  "Circuit 5 Power", id:  cir5, filters: [ *moving_avg, *pos, multiply: 1.0845 ] } }`

**Comment by flaviut** [comment, marked as answer]
> Should I be putting a multiplier in each line of ct_clamps?

Yes, exactly.

You might find it easier [to use the `calibrate_linear` filter](https://esphome.io/components/sensor/index.html?highlight=filter#calibrate-linear) to save yourself some manual math:

```
 - { phase_id: phase_a, input:  "5", power: { name:  "Circuit 5 Power", id:  cir5, filters: [ *moving_avg, *pos, calibrate_linear: ["0 -> 0", "1316 -> 1375"] ] } }
```

**Reply by Overwatching** [reply]
#### That works for me.

_I'm trying not to get overwhelmed by all the extra stuff I have to learn and keep track of at the moment.
(I have a Proxmox HA cluster set up, a second testing one, and a bunch of docker nodes. But Home Assistant and all the yaml stuff is still new to me.)
This is my first real thing I'm adding after a simple esp32+dht11 sensor that I threw together from the parts drawer, and 4x power plugs which used to have esp32s but are now manufactured with beken chips. Learning too many things at once is a recipe for forgetting everything a week from now..._

#### Additionally, do we have a way to reset the usage on the device?
Like, get rid of all the daily usage?

I have like...30-90 minutes of power usage recordings from this 1400W heater on all 16 circuits (some double wrapped since I'll be using them on 240v circuits).
That is a lot of garbage data to bee keeping on the device.

I'm assuming I should also be doing this when I stand up my "production" Home Assistant server this week? (It's on my Lab server now)

**Reply by flaviut** [reply]
> I'm trying not to get overwhelmed by all the extra stuff I have to learn and keep track of at the moment.

Haha sure, no need to get into too much all at once :)

HASS/ESPHome is a real pleasure once you get it figured out

> Additionally, do we have a way to reset the usage on the device?

On the device? Just reboot, the daily energy is only stored in RAM. I can't remember if there's a software way to reboot, but you can always flip the breakers that power it.

Also, the daily energy reading resets automatically at midnight. If you want long-term data, you'll need to set up Home Assistant to capture and store it.

**Reply by Overwatching** [reply]
I'm getting errors on `calibrate_linear` using the syntax you posted.

`... filters: [ *moving_avg, *pos, calibrate_linear: {0: 0, 1404: 1376} ] ...`

- `'from' is a required option for [calibrate_linear].`
- `'to' is a required option for [calibrate_linear].`
- `[0] is an invalid option for [calibrate_linear]. Please check the indentation.`
- `[1404] is an invalid option for [calibrate_linear]. Please check the indentation.`

I'm not super comfortable looking at yaml yet, and the esphome documentation shows the usage in multi-line format.
I'm not sure what is actually wrong.

**Reply by flaviut** [reply]
Apparently I don't understand YAML syntax either :)

I've corrected the OP.

**Reply by Overwatching** [reply]
Changed

- `calibrate_linear: {0: 0, 1405: 1376}`
to
- `calibrate_linear: [0 -> 0, 1405 -> 1376]`

and it worked.

**Reply by Overwatching** [reply]
This heating fan has a 1/2 setting too.

I wonder if I'm better off using `calibrate_polynomial` since I can get 0, half , and full power measurements.

**Reply by flaviut** [reply]
I believe the physical process here is linear, and using `calibrate_polynomial` will result in overfitting. You can pass 3+ values to `calibrate_linear`, and the system will perform a line of best fit on them.

**Reply by Overwatching** [reply]
perfect, I'll try that.

**Reply by Overwatching** [reply]
Note:

The energy values do not seem to be stored in RAM.
I left the device unplugged for 25+ minutes while I ate, and it still has the same values when it started back up.

Which leaves me with questions.
How much is getting written to the flash, and how often?
How quickly is this going to wear out the flash?

I guess I'll look through the code and see how much I understand.
Then I'll post a new discussion.
